File: pkg/msg/channel.go
package msg

import (
	"sync"
	"time"
)

// Transport exchanges raw messages with a device or peer.
type Transport interface {
	Read() ([]byte, error)
	Write([]byte) error
	Close()
}

// Channel routes raw transport traffic to subscribers based on session ownership.
type Channel struct {
	tr       Transport
	dev      Device
	width    int
	done     chan struct{}
	once     sync.Once
	mu       sync.Mutex
	queues   map[Queue]struct{}
	opening  map[string]Queue
	sessions map[uint16]Queue
	closing  map[uint16]Queue
}

const dispatchTimeout = time.Second

// NewChannel wraps the provided transport in a session-aware channel.
func NewChannel(tr Transport, dev Device, width int) *Channel {
	ch := &Channel{
		tr:       tr,
		dev:      dev,
		width:    width,
		done:     make(chan struct{}),
		queues:   make(map[Queue]struct{}),
		opening:  make(map[string]Queue),
		sessions: make(map[uint16]Queue),
		closing:  make(map[uint16]Queue),
	}

	go ch.reader()

	return ch
}

// Width returns the maximum number of inflight messages.
func (c *Channel) Width() int {
	return c.width
}

// Device returns the underlying device if available.
func (c *Channel) Device() Device {
	return c.dev
}

// Subscribe registers a queue to receive incoming messages.
func (c *Channel) Subscribe(queue Queue) {
	c.mu.Lock()
	defer c.mu.Unlock()

	// add queue
	c.queues[queue] = struct{}{}
}

// Unsubscribe removes a queue from the subscriber list.
func (c *Channel) Unsubscribe(queue Queue) {
	c.mu.Lock()
	defer c.mu.Unlock()

	// remove queue
	delete(c.queues, queue)
	for handle, owner := range c.opening {
		if owner == queue {
			delete(c.opening, handle)
		}
	}
	for session, owner := range c.sessions {
		if owner == queue {
			delete(c.sessions, session)
			delete(c.closing, session)
		}
	}
	for session, owner := range c.closing {
		if owner == queue {
			delete(c.closing, session)
		}
	}
}

// Write sends a single framed message on behalf of a specific queue. A nil
// queue means the write has no subscriber ownership context.
func (c *Channel) Write(queue Queue, msg Message) error {
	return c.write(queue, msg)
}

// Close closes the channel and the underlying transport.
func (c *Channel) Close() {
	c.closeDone()
	c.tr.Close()
}

// Done is closed when the underlying transport reader exits.
func (c *Channel) Done() <-chan struct{} {
	return c.done
}

func (c *Channel) closeDone() {
	c.once.Do(func() {
		close(c.done)
	})
}

func (c *Channel) reader() {
	defer c.closeDone()

	for {
		// read data
		data, err := c.tr.Read()
		if err != nil {
			return
		}

		// parse message
		msg, ok := Parse(data)
		if !ok {
			continue
		}

		// match targets for message
		targets := c.match(msg)

		// queue message
		for _, queue := range targets {
			select {
			case queue <- msg:
			case <-time.After(dispatchTimeout):
			}
		}
	}
}

func (c *Channel) write(from Queue, msg Message) error {
	// forward broadcast messages
	if from == nil {
		return c.tr.Write(msg.Build())
	}

	// check owned messages
	if msg.Session != 0 {
		c.mu.Lock()
		owner := c.sessions[msg.Session]
		c.mu.Unlock()
		if owner != nil && owner != from {
			return SessionWrongOwner
		}
	}

	// register write (session opens and closes)
	c.mu.Lock()
	if msg.Session == 0 && msg.Endpoint == 0x0 {
		c.opening[string(msg.Data)] = from
	}
	if msg.Session != 0 && msg.Endpoint == 0xFF {
		c.closing[msg.Session] = from
	}
	c.mu.Unlock()

	// write message
	err := c.tr.Write(msg.Build())
	if err != nil {
		// revert registrations (session opens and closes)
		c.mu.Lock()
		if msg.Session == 0 && msg.Endpoint == 0x0 && c.opening[string(msg.Data)] == from {
			delete(c.opening, string(msg.Data))
		}
		if msg.Session != 0 && msg.Endpoint == 0xFF && c.closing[msg.Session] == from {
			delete(c.closing, msg.Session)
		}
		c.mu.Unlock()

		return err
	}

	return nil
}

func (c *Channel) match(msg Message) []Queue {
	// acquire mutex
	c.mu.Lock()
	defer c.mu.Unlock()

	// handle session open replies
	if msg.Endpoint == 0x0 {
		owner := c.opening[string(msg.Data)]
		if owner != nil {
			delete(c.opening, string(msg.Data))
			_, ok := c.queues[owner]
			if ok {
				c.sessions[msg.Session] = owner
				return []Queue{owner}
			}
		}
	}

	// handle session traffic and closes
	if msg.Session != 0 {
		owner := c.sessions[msg.Session]
		if owner != nil {
			_, ok := c.queues[owner]
			if ok {
				if msg.Endpoint == 0xFF && len(msg.Data) == 0 {
					delete(c.sessions, msg.Session)
					delete(c.closing, msg.Session)
				}
				return []Queue{owner}
			}
			delete(c.sessions, msg.Session)
			delete(c.closing, msg.Session)
		}
		return nil
	}

	// handle broadcast messages
	targets := make([]Queue, 0, len(c.queues))
	for queue := range c.queues {
		targets = append(targets, queue)
	}

	return targets
}

File: web/channel.ts
import { AsyncQueue } from "./queue";

/**
 * Queue is used to receive messages from a channel.
 */
export class Queue extends AsyncQueue<Uint8Array> {}

/**
 * Channel provides the mechanism to exchange messages between a device and a client.
 */
export interface Channel {
  name(): string;
  subscribe(queue: Queue): void;
  unsubscribe(queue: Queue): void;
  write(data: Uint8Array): Promise<void>;
  close(): void;
}

/**
 * Message represents a message exchanged between a device and client.
 */
export class Message {
  session: number; // uint16
  endpoint: number; // uint8
  data: Uint8Array | null;

  constructor(session: number, endpoint: number, data: Uint8Array | null) {
    this.session = session;
    this.endpoint = endpoint;
    this.data = data;
  }

  /**
   * Returns the size of the message.
   */
  size(): number {
    return this.data?.length ?? 0;
  }
}

/**
 * Read reads a message from the queue.
 */
export async function read(queue: Queue, timeout: number): Promise<Message> {
  // read from queue
  const data = await queue.pop(timeout);
  if (!data) {
    throw new Error("timeout");
  }

  // check length and version
  if (data.length < 4 || data[0] !== 1) {
    throw new Error("invalid message");
  }

  // get view
  const view = new DataView(data.buffer);

  return new Message(
    view.getUint16(1, true),
    data[3],
    data.length > 4 ? data.slice(4) : null
  );
}

/**
 * Write writes a message to the channel.
 */
export async function write(ch: Channel, msg: Message): Promise<void> {
  // prepare data
  const data = new Uint8Array(4 + msg.size());
  const view = new DataView(data.buffer);
  view.setUint8(0, 1); // version
  view.setUint16(1, msg.session, true);
  view.setUint8(3, msg.endpoint);
  if (msg.data) {
    data.set(msg.data, 4);
  }

  // write data
  await ch.write(data);
}
